### PR TITLE
Fix "Pivot is not correctly remembered"

### LINKED ISSUE #9667: Pivot is not correctly remembered
To reproduce:

* Eg go to "Visitors => Custom variables"
* Select "Pivot report" via cog icon
* Open any other report
* Go back to "Visitors => Custom variables"
* It says the report is pivoted but it is not

Problem: The `HtmlTable` visualization is applied instead of `PivotBy` because there is no URL parameter that says "pivot is used" here: https://github.com/piwik/piwik/blob/2.16.0-rc1/plugins/CoreVisualizations/CoreVisualizations.php#L46-L50

The decision which visualization to use is made later by saved ViewDataTable parameters but we have to decide which of those two visualizations we want to use earlier. To solve this we need to give `Pivot` visualization a unique name (Visualization ID). This sounds easy but it is not that trivial since we somehow still need to disable `HtmlTable` visualization in such a case. There must be better solutions :)

### PR DESCRIPTION
**Pull request by sgiehl**
Moving all logic from the separate `PivotBy` viz to the `HtmlTable` viz solves the problem.

fixes #9667 

might also be a solution for #9716

### CODE REVIEW

File: plugins/CoreVisualizations/CoreVisualizations.php [hunk 1]
```diff
@@ -39,23 +38,6 @@ public function deleteUser($userLogin)
         ViewDataTableManager::clearUserViewDataTableParameters($userLogin);
     }
 
-    public function addViewDataTable(&$viewDataTable)
-    {
-        // Both are the same HtmlTable, just the Pivot one has some extra logic in case Pivot is used. 
-        // We don't want to use the same HtmlTable twice in the UI. Therefore we always need to remove one.
-        if (Common::getRequestVar('pivotBy', '')) {
```
Comment (sgiehl) [inline plugins/CoreVisualizations/CoreVisualizations.php]: This actually caused the problem, as for an initially loaded report no request vars are set.

File: plugins/Events/Events.php [hunk 2]
```diff
@@ -143,7 +143,8 @@ public function configureViewDataTable(ViewDataTable $view)
         $secondaryDimension = $this->getSecondaryDimensionFromRequest();
         $view->config->subtable_controller_action = API::getInstance()->getActionToLoadSubtables($apiMethod, $secondaryDimension);
 
-        if (Common::getRequestVar('pivotBy', false) === false) {
+        $pivotBy = Common::getRequestVar('pivotBy', false);
+        if (empty($pivotBy)) {
```
Comment (sgiehl) [inline plugins/Events/Events.php]: When unsetting the pivot `pivotBy`is set to the string `0`. This wasn't catched here, so the displayed columns were incorrect in this case.

File: plugins/CoreVisualizations/Visualizations/HtmlTable/AllColumns.php [hunk 3]
```diff
@@ -64,4 +64,14 @@ public function beforeGenericFiltersAreAppliedToLoadedDataTable()
             $properties->columns_to_display = $columnsToDisplay;
         });
     }
+
+    public function beforeLoadDataTable()
+    {
+        $this->requestConfig->request_parameters_to_modify['pivotBy'] = null; // always unset pivotBy
```
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/HtmlTable/AllColumns.php]: this fixes another issue that was occurring when switching all columns view after pivot was activated.

**Comment by tsteur** [comment]
Have you seen the 2 failing UI tests? http://builds-artifacts.piwik.org/piwik/piwik/pivot_9667/18398/PivotByDimension_pivoted_columns_report and http://builds-artifacts.piwik.org/piwik/piwik/pivot_9667/18398/PivotByDimension_pivoted ? I tried to find a CSS selector that matches `pivotby` dataTable but couldn't really find anything but there are now some CSS classes matching like `dataTableVizHtmlTable` etc. I presume we can still merge and it is expected. :+1:

**Comment by sgiehl** [comment]
@tsteur Yes, the small changes in the UI tests are because the css class of the datatable changed. But it is somehow expected.

**Comment by mattab** [comment]
Good find @sgiehl ! :+1:

**Pull request merged**